url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/HW4Main.java

```diff
@@ -0,0 +1,318 @@
+		long elapsed = System.currentTimeMillis() - start;
+		//System.out.println("ELAPSED TIME: " + elapsed + " milliseconds to run search");
+		return found.value;
+	}
+	
+	
+	/**
+	 * mapcolor test from devin
+	 */
+	public static void test() {
+		
+		String variables = "WA.NT.SA.NSW.Q.V";
+		String values="r.g.b";
+		String constraint = "r.g:r.b:g.r:g.b:b.r:b.g";
+		
+		
+		CSP mapcolor = new CSP(variables, values);
+		
+		mapcolor.addConstraint("WA", "NT", constraint);
+		mapcolor.addConstraint("SA", "NT", constraint);
+		mapcolor.addConstraint("SA", "WA", constraint);
+		mapcolor.addConstraint("SA", "NSW", constraint);
+		mapcolor.addConstraint("SA", "V", constraint);
+		mapcolor.addConstraint("SA", "Q", constraint);
+		mapcolor.addConstraint("NSW", "Q", constraint);
+		mapcolor.addConstraint("NSW", "V", constraint);
+		
+		PartialAssignment pa = new PartialAssignment(6);
+		
+		mapcolor.backtrackingSearch(pa);
+	}
+	
+	
+	/**
+	 * use reflection to pass a converter to print the domainlist
+	 * mostly just my effort to learn a bit about java reflection
+	 */
+	@SuppressWarnings("unchecked")
+	public static void printDL(DomainList dl, CircuitProblem cp)
+	{
+		try
+		{
+			Class cls = Class.forName("hw4.Rect");
+			Object foo = cls.newInstance();
+			Method method = cls.getMethod("intToString", new Class[] { int.class, int.class });
+			System.out.println(dl.toString(foo, method, cp.width));
+		}
+		catch (Exception e)
+		{ 
+			System.out.println("error printing: " + e);
+		}
+	}
+	
+	/**
+	 * takes a start rect to indicate left upper corner coordinates
+	 * and then given the rect r builds a list of all absolute internal coordinates
+	 * @param orig
+	 * @param cols
+	 * @param r
+	 * @return set of coordinates
+	 */
+	public static Set<Integer> coordSet(int orig, int cols, Rect r)
+	{
+		Rect o = Rect.fromInt(orig, cols);
+		int xorg = o.width;
+		int yorg = o.height;
+		
+		Set<Integer> s = new HashSet<Integer>();
+				
+		for (int x=xorg; x < (xorg + r.width); x++)
+		{
+			for (int y=yorg; y < (yorg + r.height); y++)
+			{
+				s.add(Rect.getInt(x, y, cols));
+			}
+		}
+		return s;
+	}
+	
+	/**
+	 * prints out the solution on a grid
+	 * @param m
+	 * @param cp
+	 */
+	public static void printSolution(Map<String, String> m, CircuitProblem cp)
+	{
+		if (m ==null) {return;}
+		
+		Map<String,Set<Integer>> all = new HashMap<String,Set<Integer>>();
+		Map<String, String> r = new HashMap<String, String>();
+		
+		for (String k : m.keySet())
+		{
+			r.put(m.get(k), k);
+		}
+		
+		for (String loc : m.values())
+		{
+			int l = Integer.parseInt(loc);
+			Rect i = cp.compList.get(names.indexOf(r.get(loc)));
+			all.put(r.get(loc), coordSet(l, cp.width, i));
+		}
+		
+		System.out.println();
+		boolean print = false;
+		for (int y = 0; y < cp.height; y++)
+		{
+			for (int x = 0; x < cp.width; x++)
+			{
+				for (String var : all.keySet())
+				{
+					if (all.get(var).contains(new Rect(x, y).toInt(cp.width)))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.